cpp/ql/src/experimental/Security/CWE/CWE-552/DangerousUseRename.cpp

```diff
@@ -0,0 +1,13 @@
+
+...
+  if (rename(from,to)==0) // BAD
+    return;
+  f1 = fopen(from, "r");
+  f2 = fopen(to, "w");
+...
+  if (rename(from,to)==0) // GOOD
+    return;
+  f1 = fopen(from, "r");
+  fd = open(to, O_WRONLY|O_CREAT|to_oflags, S_IRUSR|S_IWUSR);
+  f2 = fdopen(fd, "w");
+...
```

Comment:
In both case you seem to check for the proper operation of `rename` before you do anything else. Are the examples correct?

Reply:
if `rename` did not work, then in the first example we simply use unrestricted access to files, and in the second we restrict it with flags.

Reply:
I see now. This needs to be explained in the `qhelp` file. Also would to be possible to make the examples more focussed? Having `f1 = fopen(from, "r");` in there confuses me, and it's not clear to me why it's there at all.

Given the examples, it also seems that in your query you combine two cases that should really be separate:

1. `rename` fails and it followed by an `fopen` call.
2. the result of `rename` does not matter and an `open` call follows with an appropriate set of `mode` flags.

Reply:
looks really confusing.

I only work with the first situation.
`rename` failed and an attempt was made to rename the path to an open file without permissions.

how do you look at such a change `qhelp`

```
If the rename function did not work correctly, attempting to overwrite the target file may result in third-party files being affected, for example through the linking mechanism. Use access to the target file with permission setting.
If the rename function did not work correctly, attempting to overwrite the target file may result in third-party files being affected, for example through the linking mechanism. Use access to the target file with permission setting.
```